### ISSUE TITLE
What does "choosing among all families" actually mean?

### ISSUE THREAD

**Issue by scharch** [psathyrella/selection-metric-comments#8]
Going back through this, I realized that I don't fully understand what the goals here are or how you simulated the scenario (it's not explicitly in methods, that I see).

The closest I see to explaining the goals is the sentence 
>Thus while `aa-cdist` is effective at choosing the highest affinity antibodies both
> from among all families andwithin each family, since specificity is typically important, 
> it will often be necessary to combine it with such a non-sequence-based method.

If I understand correctly, then, success is defined as picking out the most "optimized" antibody(s) in a repertoire, whatever their cognate antigen. I think it would be nice if this were spelled out more carefully earlier in the paper, but so far so good. The problem to me is that it seems extraordinarily contingent on a bunch of parameters different from the ones you've scanned (at least in the ms; I'll cop to not having gone through the data in zenodo).

I'm assuming the basics of the simulation are simple enough: instead of starting with one naive sequence, you combine the final sequences sampled from the (independent) simulation a bunch of different naive sequences. But how many lineages were pooled into a "repertoire"? What was the size distribution of the lineages in the repertoire? Was `aa_cdist` (or actual simulated affinity) correlated with the number of sequences selected into the repertoire? What's the minimum size of a lineage in the repertoire for it to be considered in these calculations? Do you have any intuitions or way to simulate the impact of lineages competing with each other for the same antigen?

All of those questions lead me back to the first one, which is what do you see as the use case here? It seems that it's _not_, after all, looking at a repertoire. I guess maybe, having identified 5 anti-HA lineages experimentally, to pick the best antibody(s) from among all of those? I still have questions about how relative sizes and sampling strategies impact the way sequences in different lineages score, though...

**Comment by psathyrella** [comment]
I agree the among-families thing is a little weird, mostly because we're saying "choose from all families in the repertoire" but we also emphasize we can only know about affinity, not specificity. Maybe we should say "all families of a given specificity".

It's also worth mentioning that I started working on the among-families problem long after the within-families one, and was expecting that it would *not* work very well, so was particularly surprised at how well aa-cdist handled it (as opposed to e.g. lbi which did much better within than among). I still wish I understood more about this -- for instance I thought "amino acid distance from family consensus sequence to naive sequence" (i.e. how far the family as a whole has evolved, see table 2 of dtr input variables) would be necessary for good among-families performance, hence why I sunk so much time into the dtr. It bothers me that I don't have a better intuitive understanding for why aa-cdist does so well choosing among all families.

```having identified 5 anti-HA lineages experimentally, to pick the best antibody(s) from among all of those?```

But yes, I think that would be  a typical use case. There's a bunch of non-sequence-based ways to enrich for families of a particular specificity mentioned in the Discussion, but for instance with longitudinal sampling at day 0 and 7 after vaccination, families that expand rapidly from day 0 to 7 will be mostly vaccine specific.

As to how the simulation works now, it just simulates each family entirely separately, and mashes together as many families as desired into a "repertoire". This is in the Fig 1 caption, but maybe it isn't anywhere else (you're right it should be): "Each point shows the mean ± standard error of 30 samples, where each sample consists of 1500 sequences from 10 families." So for these "slice" plots, all the parameters are the same for each family in a repertoire, but how each family turns out varies wildly, since each family's history is determined by a very small, and thus highly stochastic, number of advantageous mutations (like 3-5). I just ran a quick test to make 10 events all with the same parameters, this family only had one advantageous mutation (top left plot has one entry to left of 0):

![p](https://user-images.githubusercontent.com/2960243/82702661-45631380-9c27-11ea-88ac-22a0f7e594de.png)

While this one had five:

![p](https://user-images.githubusercontent.com/2960243/82702715-6461a580-9c27-11ea-94a1-b0ac80b7d812.png)
Also, if selection strength were less than 1, you'd start getting a lot of families that don't even evolve toward the target sequence at all. 

So "choosing among families" would mean lumping together the sequences from the two families, and seeing if you chose the sequences with the highest affinity (in this case these'd be the ones near Kd 75 from the second family). In a scatter plot this lumping together just means throwing all the families together into one plot (ok this is all 10 families now):

![p](https://user-images.githubusercontent.com/2960243/82705097-d63bee00-9c2b-11ea-8fa3-46e23c710bf2.png)

But as I emphasize in the paper, this does *not* mean that it's a good idea to try to choose families as a whole -- I tried, and previous work has tried this approach, but there's way too much variation in affinity within each family to make this a worthwhile step. 

Now the next step is to draw different parameters for each family in the "repertoire". The only samples like this that are described in the paper were used for dtr training:

![dtr-train-samples-text](https://user-images.githubusercontent.com/2960243/82702811-92df8080-9c27-11ea-8ffd-849b853c7f42.png)
![dtr-train-samples-table](https://user-images.githubusercontent.com/2960243/82702816-94a94400-9c27-11ea-8c56-84df153b2586.png)

Now you would think (well I thought) that the among-families performance would be much worse, and qualitatively different for different metrics, for the latter case. But it's not. What I decided this meant was that the variation between families just due to random chance (getting lucky and having more advantageous mutations) even when all families have the same parameters is more important than varying the parameters between families. I.e. a family with selection strength 1 can easily get unlucky and end up with worse affinities than a family with selection strength 0.5.

But you're totally right that this ignores competition between families! This is on purpose, largely because different people have different and very strong opinions about how strong that competition is, and we didn't want to start a fight. In fact this is a paragraph from the .tex that's commented a.t.m.:

"""
%% Another limitation of our current approach is that while it can model a single family evolving toward multiple target sequences, it cannot model competition between families.                                                                                                                                                                                                                                                                                                                                              
%% Doing this would be much more computationally demanding, but would be feasible for small numbers of families.                                                                                                                                                                                                                                                                                                                                                                                                               
%% The biggest problem stemming from this limitation is that we cannot evaluate the effectiveness of using clonal family size as a method of choosing high-affinity families, since each family that we simulate operates under a fixed carrying capacity.                                                                                                                                                                                                                                                                     
%% As a practical matter, however, this may not be a significant problem, since the extent to which family size is an effective predictor depends entirely on how much competition between families really occurs.                                                                                                                                                                                                                                                                                                             
%% If different families are in perfect competition (for instance if we believe that there are either many families within each GC, or lots of transport between different GCs), then family size is by definition that family's fitness.                                                                                                                                                                                                                                                                                      
%% On the other hand if there is little between-families competition, family size would tell us nothing about fitness, and might instead be determined simply by which family happened to develop in the most well-resourced GC.                                                                                                                                                                                                                                                                                               
%% The extent to which different families are in competition would thus be an input parameter to the simulation that we would set based on experimental values, and we know without running the simulation what it would tell us at the extremes of that parameter.                                        
"""

I think simulating competition between families would be handled the same way competition within families is -- effectively by changes in the carrying capacity. If a new family in your GC gets a good mutation, then they'll start squeezing you out, which means you'll see fewer Ags, i.e. effectively lower carrying capacity. The math doing this is from the bcr-phylo paper.

Personally I think there's a fair bit of competition between families (I've seen a number of papers showing that Abs migrate between different GCs pretty freely), but, quantitatively, who knows how much.

I don't think that really answers all your questions, but I've already written too much, so just ask the unclear bits again if I missed something.
